### PR TITLE
Migrate Chips to Material 3

### LINKED ISSUE #99024: Update chips to support Material 3
image

As part of the migration to [Material Design 3](https://m3.material.io/), the `Chip` widgets need to be updated to the latest design spec:

https://m3.material.io/components/chips/overview

### PR DESCRIPTION
**Pull request by darrenaustin**
Part of: https://github.com/flutter/flutter/issues/91605

Updated the chip widgets `ActionChip`, `FilterChip`, `ChoiceChip` and `InputChip`  with support for [Material Design 3](https://m3.material.io/components/chips/overview).

M3 Chip
M3 Chip dark

In order to use the chip widgets with the new Material 3 defaults, turn on the `useMaterial3` flag in the `ThemeData`:

```dart
  return MaterialApp(
    theme: ThemeData(useMaterial3: true),
    // ...
  );
```

The names of some of the chips in Material Design 3 are a little different from the previous version. You can use the following to implement M3 designs:

- Assist chip: `ActionChip`
- Filter chip: `FilterChip` or `ChoiceChip` (for single selection)
- Input chip: `InputChip`
- Suggestion chip: `ActionChip`

Fixes: #99024

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/material/chip_action.dart [hunk 1]
```diff
@@ -145,12 +165,87 @@ class ActionChip extends StatelessWidget implements ChipAttributes, TappableChip
...
+// Generated code to the end of this file. Do not edit by hand.
+// These defaults are generated from the Material Design Token
+// database by the script dev/tools/gen_defaults/bin/gen_defaults.dart.
+
+// Generated version v0_101
+class _TokenDefaultsM3 extends ChipThemeData {
+  const _TokenDefaultsM3(this.context, this.isEnabled)
+    : super(
+        elevation: 0.0,
+        shape: const RoundedRectangleBorder(borderRadius: BorderRadius.only(topLeft: Radius.circular(8.0), topRight: Radius.circular(8.0), bottomLeft: Radius.circular(8.0), bottomRight: Radius.circular(8.0))),
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/chip_action.dart:192]: Could be `BorderRadius.all(Radius.circular(8))`

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/chip.dart [hunk 2]
```diff
@@ -1247,6 +1310,9 @@ class _RawChipState extends State<RawChip> with MaterialStateMixin, TickerProvid
         child: result,
       ),
     );
+    // if (height != null) {
```
[the remark this answers is not in the document]
Comment (darrenaustin) [reply packages/flutter/lib/src/material/chip.dart:1313]: Yup, I had deleted it, but perhaps it snuck back in with a merge I was doing.

File: packages/flutter/lib/src/material/chip_action.dart (hunk 1 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/chip_action.dart:192]: Yup, this is a generator problem. One that will be fixed with the BottomAppBar PR.

**Comment by rydmike** [comment]
@darrenaustin / @HansMuller Is it on purpose that plain just `Chip` does not get any M3 styling when opting in on `useMaterial3`? It keeps the M2 style:

Screen Shot 2022-07-19 at 18 33 47

Seems a bit inconsistent that it does not default to the M3 styling as well when M3 is used. Is it on purpose or an oversight, or just not done yet? In any case, if left like this, I'm sure users will wonder about it and stumble over this inconsistency.

**Comment by darrenaustin** [comment]
It is more an oversight than anything else. The M3 spec doesn't describe a 'plain' chip, so i tis not obvious what "M3 styling" should apply to Flutter's generic `Chip` widget. That said, we could probably reuse the Assist chip defaults for the vanilla Chip widget.

**Comment by rydmike** [comment]
Thanks @darrenaustin, looking at M3 spec I figured that might be the background too. Using same defaults as AssistChip sounds good. Even if not mentioned in M3 spec I think it would make sense in Flutter from a design consistency viewpoint.

**Pull request merged**